pkg/controller/awsmachine/awsmachine_controller.go

```diff
@@ -69,18 +103,283 @@ type ReconcileAWSMachine struct {
+	exist, err := r.exists(scope)
+	if err != nil {
+		klog.Errorf("Failed to check if Machine %q infrastructure exists: %v", scope.Name(), err)
+		return err
+	}
+
+	// Reconcile ProviderID.
+	if pid := scope.GetProviderID(); pid == nil || *pid == "" {
+		scope.SetProviderID(fmt.Sprintf("aws:////%s", *scope.GetInstanceID()))
+	}
+
+	if exist {
+		scope.Info("Reconciling Machine triggers idempotent update")
+		return r.update(scope)
+	}
+
+	// Machine resource created. Machine does not yet exist.
+	scope.Info("Reconciling Machine triggers idempotent create")
+	return r.create(scope)
+}
+
+// create creates a machine and is invoked by the machine controller.
+func (r *ReconcileAWSMachine) create(scope *actuators.MachineScope) error {
+	if scope.Parent.Cluster.Annotations[infrav1.AnnotationClusterInfrastructureReady] != infrav1.ValueReady {
+		scope.Info("Cluster infrastructure is not ready yet - requeuing machine")
+		return &capierrors.RequeueAfterError{RequeueAfter: waitForClusterInfrastructureReadyDuration}
+	}
+
+	ec2svc := ec2.NewService(scope.Parent)
+
+	scope.Info("Retrieving machines for cluster")
+	machineList := &clusterv1.MachineList{}
+	if err := r.List(context.Background(), machineList, actuators.ListOptionsForCluster(scope.Parent.Name())); err != nil {
+		return errors.Wrapf(err, "failed to retrieve machines in cluster %q", scope.Parent.Name())
+	}
+
+	controlPlaneMachines := util.GetControlPlaneMachinesFromList(machineList)
+	if len(controlPlaneMachines) == 0 {
+		scope.Info("No control plane machines exist yet - requeuing")
+		return &capierrors.RequeueAfterError{RequeueAfter: waitForControlPlaneMachineExistenceDuration}
+	}
+
+	// Create the Machine.
+	i, err := ec2svc.CreateOrGetMachine(scope)
+	if err != nil {
+		return errors.Errorf("failed to create or get machine: %+v", err)
+	}
+
+	scope.SetInstanceID(i.ID)
+	scope.SetInstanceState(i.State)
+	scope.SetAnnotation("cluster-api-provider-aws", "true")
+
+	if err := r.reconcileLBAttachment(scope, i); err != nil {
+		return errors.Errorf("failed to reconcile LB attachment: %+v", err)
+	}
+
+	scope.Info("Create completed")
+	return nil
+}
+
+func (r *ReconcileAWSMachine) exists(scope *actuators.MachineScope) (bool, error) {
+	ec2svc := ec2.NewService(scope.Parent)
+
+	// TODO worry about pointers. instance if exists returns *any* instance
+	if scope.GetInstanceID() == nil {
+		return false, nil
+	}
+
+	instance, err := ec2svc.InstanceIfExists(scope.GetInstanceID())
+	if err != nil {
+		return false, errors.Errorf("failed to retrieve instance: %+v", err)
+	}
+
+	if instance == nil {
+		return false, nil
+	}
+
+	scope.Info("Found instance for machine", "instance", instance)
+
+	switch instance.State {
+	case infrav1.InstanceStateRunning:
+		scope.Info("Machine instance is running", "instance-id", *scope.GetInstanceID())
+	case infrav1.InstanceStatePending:
+		scope.Info("Machine instance is pending", "instance-id", *scope.GetInstanceID())
```

Comment:
We said that the Machine should enter an error state if the underlying instance is no longer available, should that be taken into account here?

Reply:
I'll follow-up with more logic changes in later PRs, how does that sound? There is a bunch of things to take care of, e.g. Ready/Addresses/etc